contracts/randdrop/src/contract.rs

```diff
@@ -0,0 +1,763 @@
+        .map_err(|_| ContractError::InvalidRandomness)?;
+
+    // extract participant address
+    let job_id = callback.job_id;
+    let participant_address = job_id
+        .strip_prefix("randdrop-")
+        .unwrap_or_else(|| panic!("Strange, how is the job-id not prefixed with randdrop-"));
+    let participant_address = deps.api.addr_validate(participant_address)?;
+
+    // Make sure the participant is registered
+    let participant_data = PARTICIPANTS.load(deps.storage, &participant_address)?;
+    if participant_data.nois_randomness.is_some() || participant_data.is_winner.is_some() {
+        panic!("Strange, participant's randomness already received")
+    }
+    let is_winner = is_randdrop_winner(&info.sender, randomness);
+    PARTICIPANTS.save(
+        deps.storage,
+        &participant_address,
+        &ParticipantData {
+            nois_randomness: Some(randomness),
+            randdrop_amount: participant_data.randdrop_amount,
+            is_winner: Some(is_winner),
+        },
+    )?;
+
+    Ok(Response::new()
+        .add_attribute("action", "receive-randomness")
+        .add_attribute("job_id", job_id)
+        .add_attribute("participant", participant_address)
+        .add_attribute("randdrop_amount", participant_data.randdrop_amount)
+        .add_attribute("is_winner", is_winner.to_string()))
+}
+
+fn execute_withdraw_all(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    address: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    // check the calling address is the authorised address
+    ensure_eq!(info.sender, config.manager, ContractError::Unauthorized);
+
+    let mut out_msgs = Vec::<BankMsg>::new();
+
+    let amount = deps
+        .querier
+        .query_balance(env.contract.address.clone(), config.randdrop_denom.clone())?;
+    let msg = BankMsg::Send {
+        to_address: address.clone(),
+        amount: vec![amount],
+    };
+    out_msgs.push(msg);
+    // If the tokens to pay the proxy are different from the randropped tokens then withdraw those as well
+    if config.randdrop_denom != config.nois_proxy.price.denom {
+        let amount = deps
+            .querier
+            .query_balance(env.contract.address, config.nois_proxy.price.denom)?;
+        let msg = BankMsg::Send {
+            to_address: address,
+            amount: vec![amount],
+        };
+        out_msgs.push(msg);
+    }
+
+    let res = Response::new()
+        .add_messages(out_msgs)
+        .add_attribute("action", "withdraw_all");
+    Ok(res)
+}
+
+fn is_randdrop_winner(sender: &Addr, randomness: [u8; 32]) -> bool {
+    let sender_hash: [u8; 32] = Sha256::digest(sender.as_bytes()).into();
+
+    // Hash the combined value using SHA256 to generate a random number between 1 and 3
+    let mut hasher = Sha256::new();
+    // Concatenate the randomness and sender hash values
+    hasher.update(randomness);
+    hasher.update(sender_hash);
```

Comment:
We can just hash the sender in here, then we don't need two hashes:

```suggestion
    hasher.update(sender.as_bytes());
```